### ISSUE TITLE
Absolute URL check should be a part of fetch() but not Request class

### ISSUE THREAD

**Issue by kirill-konshin** [bitinn/node-fetch#43]
It should be possible to create an instance of `Request` with any URL (non-absolute or non http(s) protocol) because it's just a container, and such constraint is not the responsibility of `Request` class:

```js
var req = new Request('foo'); // this is legit
```

Absolute URLs constraint is a responsibility of `fetch()` function, as it can't send such `Requests`, but `Requests` themselves are valid.

**Comment by bitinn** [comment]
The best I can say at the moment: try not to have this problem, design around it.

It sucks, but it would also be safer because `fetch(full-url)` has been the main focus (more usage and likely better implemented).

For edge cases, IMO it's safer rely on your own code, until a fully-tested implementation arrives.

Nevertheless, thx for opening this issue.

**Comment by kirill-konshin** [comment]
You're right, that's mostly an edge case. The main purpose of this issue is to have a placeholder.

Architecturally having the check inside `fetch()` is a more robust solution.

To me, the code should be like:

```js
function fetch(input, init){

    // wrap everything with a promise right away to make sure no uncaught exceptions will be thrown
    return new Promise(function(resolve, reject){

        // immediately boil down to Request instance if initializers were passed
        if (!Request.isPrototypeOf(input)) { input = new Request(input, init); }

        var url_parsed = parse_url(input.url);

        if (!url_parsed.protocol || !url_parsed.hostname) {
            throw new Error('only absolute urls are supported');
	}

        ...

    });

}
```

P.S. "it's safer rely on your own code, until a fully-tested implementation arrives" -- exactly, that's the reason why we need to keep track of such non-spec-compatible issues in order to build the best implementation.

**Comment by kirill-konshin** [comment]
Btw

```js
try {
			options = new Request(url, opts);
		} catch (err) {
			reject(err);
			return;
		}
```

is not needed since anything thrown inside Promise resolver function will result in rejection.

**Comment by kirill-konshin** [comment]
Let's revisit this one. Shall I do a PR?

**Comment by bitinn** [comment]
@kirill-konshin I am ok with it, just know that cases like #75 might need additional testing.

**Comment by kirill-konshin** [comment]
I'd say that #75 should be resolved in a separate PR. And yes, I already saw that issue.

**Comment by bitinn** [comment]
Yeah, we will have to implement this before file url can be supported anyway.

But I don't currently have a use-case for this, do you?

**Comment by kirill-konshin** [comment]
The use case is for testing or to pass around pre-built Request objects. Plus for the sake of cleaner implementation. As I said earlier, this check semantically belongs to `fetch()` method, not to `Request` constructor...

**Comment by bitinn** [comment]
@kirill-konshin oh ok you mean `fetch` still enforce that, I think that's alright.

**Comment by kirill-konshin** [comment]
Yes, my entire comment https://github.com/bitinn/node-fetch/issues/43#issuecomment-139448648 was about this moving of functionality. I'll make a PR in the next few days.

**Comment by bitinn** [comment]
fixed this in v1.5.3, Request now takes any url (fetch still reject non-http url though).
